libcudacxx/include/cuda/std/__type_traits/copy_cvref.h

```diff
@@ -20,32 +20,182 @@
 #  pragma system_header
 #endif // no system header
 
-#include <cuda/std/__type_traits/add_lvalue_reference.h>
-#include <cuda/std/__type_traits/add_rvalue_reference.h>
-#include <cuda/std/__type_traits/copy_cv.h>
-
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-template <class _From, class _To>
-struct __copy_cvref
+struct __apply_cvref_
 {
-  using type = __copy_cv_t<_From, _To>;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = _Tp;
 };
 
-template <class _From, class _To>
-struct __copy_cvref<_From&, _To>
+struct __apply_cvref_c
 {
-  using type = __add_lvalue_reference_t<__copy_cv_t<_From, _To>>;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const _Tp;
 };
 
-template <class _From, class _To>
-struct __copy_cvref<_From&&, _To>
+struct __apply_cvref_v
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = volatile _Tp;
+};
+
+struct __apply_cvref_cv
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const volatile _Tp;
+};
+
+struct __apply_cvref_lr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = _Tp&;
+};
+
+struct __apply_cvref_clr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const _Tp&;
+};
+
+struct __apply_cvref_vlr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = volatile _Tp&;
+};
+
+struct __apply_cvref_cvlr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const volatile _Tp&;
+};
+
+struct __apply_cvref_rr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = _Tp&&;
+};
+
+struct __apply_cvref_crr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const _Tp&&;
+};
+
+struct __apply_cvref_vrr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = volatile _Tp&&;
+};
+
+struct __apply_cvref_cvrr
 {
-  using type = __add_rvalue_reference_t<__copy_cv_t<_From, _To>>;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const volatile _Tp&&;
 };
 
+#ifndef _CCCL_NO_VARIABLE_TEMPLATES
```

Comment:
This might be preexisting but why is this dependent on variable templates?

Reply:
I'm not sure what you mean by the "second part". This uses partial specializations of a variable template to pick the right metafunction before applying it. Variable templates are less expensive to instantiate than class templates.

Reply:
Oh I internally read that as a function call